src/Settings/Default.hs

```diff
@@ -30,21 +25,18 @@ import Settings.Builders.Ld
 import Settings.Builders.Make
 import Settings.Builders.RunTest
 import Settings.Builders.Xelatex
-import Settings.Packages.Base
-import Settings.Packages.Cabal
-import Settings.Packages.Compiler
-import Settings.Packages.Ghc
-import Settings.Packages.GhcCabal
-import Settings.Packages.Ghci
-import Settings.Packages.GhcPkg
-import Settings.Packages.GhcPrim
-import Settings.Packages.Haddock
-import Settings.Packages.Haskeline
-import Settings.Packages.IntegerGmp
+import Settings.Packages
```

Comment:
This doesn't seem to be fundamentally linked to the problem at hand, but is just a refactoring. I suggest to remove this refactoring from this PR, which is already huge.

Reply:
I tried "undoing" this in [this patch](https://gist.github.com/alpmestan/3ebfe3b63aad7027da23d2be9d51ca26) but I get:

```
unrecognised command: --flags= (try --help)
shakeArgsWith   0.000s    0%                           
Function shake  8.965s   56%  =========================
Database read   0.046s    0%                           
With database   0.000s    0%                           
Running rules   6.840s   43%  ===================      
Total          15.852s  100%                           
Error when running Shake build system:
* _ddd/stage1/lib/package.conf.d/Cabal-2.3.0.0.conf
* _ddd/stage1/lib/platformConstants
* _ddd/generated/platformConstants
* _ddd/stage0/bin/deriveConstants
* OracleQ (PackageDataFile (Context {stage = Stage0, package = Package {pkgLanguage = Haskell, pkgType = Program, pkgName = "deriveConstants", pkgPath = "utils/deriveConstants"}, way = v}))
* _ddd/stage0/utils/deriveConstants/setup-config
ExitFailure 1
```

Any idea of what's wrong, @snowleopard or @angerman ?

Reply:
@alpmestan If I interpret the error correctly, an unrecognised argument `--flags` was passed when configuring the `deriveConstants` package. 

This seems to be the change in question:

```diff
-    mconcat [ builder GhcCabal ? arg ("--flags=" ++ integerLibraryName)
+    mconcat [ builder CabalFlags ? arg ('+':integerLibraryName)
```

Perhaps we should keep this change?

Reply:
Well, that's the thing, my patch _removes_ `--flags`... so I'm really confused.

Reply:
I though you patch now *undos* this removal!

Reply:
OK, let me clarify what my patch does. Moritz's branch introduced this `Settings.Packages` module and stuffed all the package-specific flags and what not under a single thing. In the aforementionned patch I delete this module and just do like `master`, except that I changed some `GhcCabal`s into `GhcCabal Conf`s and some others into `CabalFlags`, as appropriate. So there's only one place where we see `--flags` now, it's in `Hadrian.Haskell.Cabal.Parse`, but... why unrecognised command?

Reply:
In this case it may be useful to look at the full command line which fails.

Reply:
What fails is
```
          Hooks.defaultMainWithHooksNoReadArgs hooks gpd (argList ++ ["--flags=" ++ unwords flagList])
```
Now as to why it fails. It fails because `flagList` **and** `argList` are empty. To compute `argList` we use
```
        argList <- interpret (target context (GhcCabal Conf stage) [] []) =<< args <$> flavour
```
this should have looked something like:
```
["configure","--cabal-file","utils/unlit/unlit.cabal","--distdir","/ghc/_build/stage1/utils/unlit","--ipid","$pkg-$version","--prefix","${pkgroot}/..","--with-ghc=/ghc/_build/stage0/bin/ghc","--with-ghc-pkg=/ghc/_build/stage0/bin/ghc-pkg","--ghc-pkg-option=--global-package-db=/ghc/_build/stage1/lib/package.conf.d","--enable-library-vanilla","--enable-library-for-ghci","--disable-library-profiling","--disable-shared","--configure-option=CFLAGS=-fno-stack-protector -I/ghc/_build/generated -I/ghc/utils/unlit -I/ghc/includes","--configure-option=CPPFLAGS=-I_build/generated","--gcc-options=-fno-stack-protector -I/ghc/_build/generated -I/ghc/utils/unlit -I/ghc/includes","--configure-option=--with-cc=/usr/bin/gcc","--ghc-option=-ghcversion-file=/ghc/_build/generated/ghcversion.h","--with-gcc=/usr/bin/gcc","--with-ld=/usr/bin/ld","--with-ar=/usr/bin/ar","--with-alex=/.cabal/bin/alex","--with-happy=/.cabal/bin/happy","-v0","--configure-option=--quiet","--configure-option=--disable-option-checking"]
```
but was `[]`.

Reply:
The reason, is likely:
```diff
     , ghcBuilderArgs
-    , ghcCabalBuilderArgs
     , ghcPkgBuilderArgs
```

Reply:
Apart from that, it seems that undoing this will need to introduce at least:
```
- Package.ByteString
- Package.Text
- Package.Haskeline
- Package.Hsc2hs
```

I'm not sure pulling them out now, just to collapse them all together into `Packages` again is very beneficial. The reason to consolidate them all in a single file was to prevent the need to create additional files for each adjustment. An added benefit is that it is much easier to understand how configurations across various packages happen.

In general the build system should be built to be hacked on easily. If I need to make a change to any package that's not yet in Hadrian, and consequently have to create a custom module first ( instead of just adding the necessary modifications to a single module, where I can also see how modifications are done to other packages), I think we are impeding rather than supporting the person.

Reply:
@angerman Personally I'm also leaning towards putting all arguments in a single file. Back when Hadrian was just starting, it was believed that having separate per-package files would be convenient, but after working on Hadrian for a long time, I only see this separation as an unnecessary hurdle and code bloat.

What I don't like is that this change is done as part of this PR and therefore appears to be intrinsically caused by it, which is not the case. In an ideal world, this would be a separate PR, on which other people would be able to comment and raise potential concerns about this decision. Also, having it as a separate PR makes it easier to revert it if need be.

I leave this up to @alpmestan: if splitting it off into a separate PR is difficult, let's merge it as is. I don't want to overcomplicate @alpmestan's task too much.

Reply:
Well, I could do with one less task to take care of =) I suggest that we keep things like they are now for this PR, and we can proceed to or just discuss potential adjustments separately. @ang

Reply:
@alpmestan Agreed!